src/soko_game.rs

```diff
@@ -72,19 +85,33 @@ pub fn update(model: &mut Model, msg: GameAction) -> Option<GameAction> {
             }
         }
         GameAction::Reset => {
-            game.history.push(game.window.world.clone());
-            if let Some(prev_level) = game.history.first() {
-                game.window.world = prev_level.clone();
-            }
+            reset_world(game);
         }
         GameAction::ZoomClose => game.window.zoom = Zoom::Close,
         GameAction::ZoomMiddle => game.window.zoom = Zoom::Middle,
         GameAction::ZoomFar => game.window.zoom = Zoom::Far,
         GameAction::None => {}
+        GameAction::Win => {}
     };
     None
 }
 
+// If you have no history, you cannot undo, and you cannot
+// revert to the past
+pub fn reset_history(game: &mut crate::types::Game) {
```

Comment:
`reset_history` and `reset_world` both feel like they should be methods on game ie.

```
impl Game {
    pub fn reset_history(self: &mut Game) {
```

It's normal to have the type `Game` defined in `types.rs` and the `impl` in a different file. Think of `types.rs` a bit like an `.h` file just in the sense that definitions go there, you can have the implementation in the `.h` but you might also have the implementation somewhere else.

Last thing is that you should import Game in the use `crate::types::{` at the top of the file rather than doing `crate::types::Game` here.

Reply:
resolved and refactored it nicely